### PR TITLE
[SPARK-30996][SQL] invaild parquet map key type

### PR DESCRIPTION
**Pull request by Le-Dai**
### What changes were proposed in this pull request?

add parquet Map Key type verification 

### Why are the changes needed?

Illegal parquet data type should not be written

### Does this PR introduce any user-facing change?

no

### How was this patch tested?

add tests

### CODE REVIEW

**Comment by MrDLontheway** [comment]
cc @maropu @dongjoon-hyun

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormatSuite.scala [hunk 1]
```diff
@@ -56,4 +55,22 @@ class ParquetFileFormatSuite extends QueryTest with ParquetTest with SharedSpark
...
+
+  test("read parquet map invaild key type") {
+    withTempDir { dir =>
+      val spark_session = spark
+      import spark_session.implicits._
+      val df = Seq(Map(Map(1 -> 1) -> 1L)).toDF("map")
+      val exception = intercept[AnalysisException] {
+        df.write.mode("overwrite").parquet(dir.getPath)
+      }
+      assert(exception.getMessage().contains("Parquet data source does not support"))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormatSuite.scala]: I'm not sure about what's your suggestion;
```
scala> val df = Seq(Map(Map(1 -> 1) -> 1L)).toDF("map")
df: org.apache.spark.sql.DataFrame = [map: map<map<int,int>,bigint>]

scala> df.printSchema
root
 |-- map: map (nullable = true)
 |    |-- key: map
 |    |    |-- key: integer
 |    |    |-- value: integer (valueContainsNull = false)
 |    |-- value: long (valueContainsNull = false)


scala> df.write.mode("overwrite").parquet("/tmp/test")

scala>  spark.read.load("/tmp/test").printSchema()
root
 |-- map: map (nullable = true)
 |    |-- key: map
 |    |    |-- key: integer
 |    |    |-- value: integer (valueContainsNull = true)
 |    |-- value: long (valueContainsNull = true)
```
Any issue here?

Comment (MrDLontheway) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormatSuite.scala]: o sorry , I seem to be wrong
This side will make it impossible to read. Let me review it again.

Seq(Seq(Map(Map(1 -> 1) -> StructValue(1L)))).toDF("long1")
root
 |-- long1: array (nullable = true)
 |    |-- element: map (containsNull = true)
 |    |    |-- key: map
 |    |    |    |-- key: integer
 |    |    |    |-- value: integer (valueContainsNull = false)
 |    |    |-- value: struct (valueContainsNull = true)
 |    |    |    |-- l: long (nullable = false)

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormatSuite.scala]: btw, to make reviwers understood easily,  you had better clearly write the PR description, what's an issue, example queries to reproduce the issue, and how-to-fix.

Comment (MrDLontheway) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormatSuite.scala]: thanks

**Comment by MrDLontheway** [comment]
https://github.com/apache/spark/blob/336621e277794ab2e6e917391a928ec662498fcf/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala#L228

The key is checked here, it seems that the keyType is not necessarily Primitive.
After I remove it, I can read the previous data normally. Can anyone explain this place?

**Pull request closed without merging**